Executor sample gRPC local loggger fails

Running the gRPC local logger sample found here:

https://github.com/SeldonIO/seldon-core/tree/master/executor/samples/local/logger

The following error occurs on the final step `make grpc_test` :

```
cd ../../../proto && grpcurl -d '{"data":{"ndarray":[[1.0,2.0]]}}' -plaintext -proto ./prediction.proto  0.0.0.0:5000 seldon.protos.Seldon/Predict
ERROR:
  Code: Unavailable
  Message: transport is closing
Makefile:32: recipe for target 'grpc_test' failed
make: *** [grpc_test] Error 1
```

The output from the local executor `make run_grpc_executor` :

```
../../../executor --sdep seldon-model --namespace default --predictor example --file ./model_grpc.yaml --http_port 8000 --grpc_port 5000 --transport grpc
{"level":"info","ts":1588148404.286985,"logger":"entrypoint","msg":"Flags","transport":"grpc"}
{"level":"info","ts":1588148404.2870219,"logger":"entrypoint","msg":"Trying to get predictor from file"}
ERRO[0000] open /etc/podinfo/annotations: no such file or directoryFailed to load annotations  source="main.go:215"
{"level":"info","ts":1588148404.2898488,"logger":"entrypoint","msg":"Starting","worker":1}
{"level":"info","ts":1588148404.289894,"logger":"entrypoint","msg":"Starting","worker":2}
{"level":"info","ts":1588148404.289902,"logger":"entrypoint","msg":"Starting","worker":3}
{"level":"info","ts":1588148404.289908,"logger":"entrypoint","msg":"Starting","worker":4}
{"level":"info","ts":1588148404.289913,"logger":"entrypoint","msg":"Starting","worker":5}
{"level":"info","ts":1588148404.2921605,"logger":"entrypoint","msg":"Running http probes only server ","port":8000}
{"level":"info","ts":1588148404.2922218,"logger":"entrypoint","msg":"Running grpc server ","port":5000}
{"level":"info","ts":1588148404.2925289,"logger":"entrypoint","msg":"Setting max message size ","size":2147483647}
{"level":"info","ts":1588148404.29388,"logger":"SeldonRestApi","msg":"Listening","Address":"0.0.0.0:8000"}
panic: interface conversion: interface {} is nil, not string

goroutine 121 [running]:
github.com/seldonio/seldon-core/executor/predictor.(*PredictorProcess).logPayload(0xc0005da730, 0xc000186df0, 0xa, 0xc000331180, 0x17fa33b, 0x7, 0x1a39220, 0xc0002ae200, 0xc0005a8420, 0xc0005a8420)
	/home/gl/tmp/multiplex/seldon-core/executor/predictor/predictor_process.go:263 +0x3f9
github.com/seldonio/seldon-core/executor/predictor.(*PredictorProcess).Predict(0xc0005da730, 0xc0000f2a00, 0x1a39220, 0xc0002ae200, 0xc000598640, 0x0, 0x0, 0x203000)
	/home/gl/tmp/multiplex/seldon-core/executor/predictor/predictor_process.go:271 +0x328
github.com/seldonio/seldon-core/executor/api/grpc/seldon.GrpcSeldonServer.Predict(0x1a72c60, 0xc0000e9040, 0xc000202120, 0x1a5e200, 0xc0002da160, 0xc000100000, 0x7ffceab1b13f, 0x7, 0x1a4f8a0, 0xc00032ff20, ...)
	/home/gl/tmp/multiplex/seldon-core/executor/api/grpc/seldon/server.go:37 +0x28e
github.com/seldonio/seldon-core/executor/api/grpc/seldon/proto._Seldon_Predict_Handler.func1(0x1a4f8a0, 0xc00032ff20, 0x1774b20, 0xc000256140, 0x181688d, 0x1d, 0x1a4f8a0, 0xc00032ff20)
	/home/gl/tmp/multiplex/seldon-core/executor/api/grpc/seldon/proto/prediction.pb.go:1538 +0x86
github.com/grpc-ecosystem/go-grpc-middleware/tracing/opentracing.UnaryServerInterceptor.func1(0x1a4f8a0, 0xc00032f620, 0x1774b20, 0xc000256140, 0xc000598520, 0xc000598580, 0x484d16, 0x5ea938f1, 0x1f39637a, 0x1a66399e06b)
	/home/gl/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.1.0/tracing/opentracing/server_interceptors.go:31 +0x165
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x1a4f8a0, 0xc00032f620, 0x1774b20, 0xc000256140, 0x20, 0x20, 0x1699760, 0xc000256140)
	/home/gl/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.1.0/chain.go:25 +0x63
github.com/seldonio/seldon-core/executor/api/metric.(*ServerMetrics).UnaryServerInterceptor.func1(0x1a4f8a0, 0xc00032f620, 0x1774b20, 0xc000256140, 0xc000598520, 0xc0005985a0, 0x12eda2a, 0x1699760, 0xc0005985c0, 0xc000598520)
	/home/gl/tmp/multiplex/seldon-core/executor/api/metric/server.go:43 +0xa0
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x1a4f8a0, 0xc00032f620, 0x1774b20, 0xc000256140, 0xc0002b6000, 0x0, 0xc0003ddac0, 0x40e408)
	/home/gl/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.1.0/chain.go:25 +0x63
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1(0x1a4f8a0, 0xc00032f620, 0x1774b20, 0xc000256140, 0xc000598520, 0xc000598580, 0xc000081b30, 0x48df88, 0x16cf060, 0xc00032f620)
	/home/gl/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.1.0/chain.go:34 +0xd5
github.com/seldonio/seldon-core/executor/api/grpc/seldon/proto._Seldon_Predict_Handler(0x15eb960, 0xc000592200, 0x1a4f8a0, 0xc00032f620, 0xc00020e180, 0xc0002b0330, 0x1a4f8a0, 0xc00032f620, 0xc000118c40, 0x1e)
	/home/gl/tmp/multiplex/seldon-core/executor/api/grpc/seldon/proto/prediction.pb.go:1540 +0x14b
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000511200, 0x1a6b2e0, 0xc000511380, 0xc0002b6000, 0xc0002b0450, 0x2692c60, 0x0, 0x0, 0x0)
	/home/gl/go/pkg/mod/google.golang.org/grpc@v1.27.0/server.go:1024 +0x4f4
google.golang.org/grpc.(*Server).handleStream(0xc000511200, 0x1a6b2e0, 0xc000511380, 0xc0002b6000, 0x0)
	/home/gl/go/pkg/mod/google.golang.org/grpc@v1.27.0/server.go:1313 +0xd97
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0000441d0, 0xc000511200, 0x1a6b2e0, 0xc000511380, 0xc0002b6000)
	/home/gl/go/pkg/mod/google.golang.org/grpc@v1.27.0/server.go:722 +0xbb
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/home/gl/go/pkg/mod/google.golang.org/grpc@v1.27.0/server.go:720 +0xa1
Makefile:26: recipe for target 'run_grpc_executor' failed
make: *** [run_grpc_executor] Error 2
```

--- a/executor/api/grpc/seldon/server.go
+++ b/executor/api/grpc/seldon/server.go
@@ -3,7 +3,6 @@ package seldon
 import (
 	"context"
 	"net/url"
-	"strings"
 
 	"github.com/go-logr/logr"
 	"github.com/seldonio/seldon-core/executor/api/client"
@@ -35,7 +34,7 @@ func NewGrpcSeldonServer(predictor *v1.PredictorSpec, client client.SeldonApiCli
 
 func (g GrpcSeldonServer) Predict(ctx context.Context, req *proto.SeldonMessage) (*proto.SeldonMessage, error) {
 	md := grpc.CollectMetadata(ctx)
-	ctx = context.WithValue(ctx, payload.SeldonPUIDHeader, md[strings.ToLower(payload.SeldonPUIDHeader)][0])
+	ctx = context.WithValue(ctx, payload.SeldonPUIDHeader, md.Get(payload.SeldonPUIDHeader)[0])
 	seldonPredictorProcess := predictor.NewPredictorProcess(ctx, g.Client, logf.Log.WithName("SeldonMessageRestClient"), g.ServerUrl, g.Namespace, md)
 	reqPayload := payload.ProtoPayload{Msg: req}
 	resPayload, err := seldonPredictorProcess.Predict(g.predictor.Graph, &reqPayload)

--- a/executor/api/grpc/seldon/server.go
+++ b/executor/api/grpc/seldon/server.go
@@ -3,6 +3,7 @@ package seldon
 import (
 	"context"
 	"net/url"
+	"strings"
 
 	"github.com/go-logr/logr"
 	"github.com/seldonio/seldon-core/executor/api/client"
@@ -34,14 +35,15 @@ func NewGrpcSeldonServer(predictor *v1.PredictorSpec, client client.SeldonApiCli
 
 func (g GrpcSeldonServer) Predict(ctx context.Context, req *proto.SeldonMessage) (*proto.SeldonMessage, error) {
 	md := grpc.CollectMetadata(ctx)
-	ctx = context.WithValue(ctx, payload.SeldonPUIDHeader, md[payload.SeldonPUIDHeader])
+	ctx = context.WithValue(ctx, payload.SeldonPUIDHeader, md[strings.ToLower(payload.SeldonPUIDHeader)][0])
 	seldonPredictorProcess := predictor.NewPredictorProcess(ctx, g.Client, logf.Log.WithName("SeldonMessageRestClient"), g.ServerUrl, g.Namespace, md)
 	reqPayload := payload.ProtoPayload{Msg: req}
 	resPayload, err := seldonPredictorProcess.Predict(g.predictor.Graph, &reqPayload)
 	if err != nil {
 		g.Log.Error(err, "Failed to call predict")
+		return payloadToMessage(resPayload), err
 	}
-	return resPayload.GetPayload().(*proto.SeldonMessage), err
+	return payloadToMessage(resPayload), nil
 }
 
 func (g GrpcSeldonServer) SendFeedback(ctx context.Context, req *proto.Feedback) (*proto.SeldonMessage, error) {
@@ -50,6 +52,14 @@ func (g GrpcSeldonServer) SendFeedback(ctx context.Context, req *proto.Feedback)
 	resPayload, err := seldonPredictorProcess.Feedback(g.predictor.Graph, &reqPayload)
 	if err != nil {
 		g.Log.Error(err, "Failed to call feedback")
+		return payloadToMessage(resPayload), err
 	}
-	return resPayload.GetPayload().(*proto.SeldonMessage), err
+	return payloadToMessage(resPayload), nil
+}
+
+func payloadToMessage(p payload.SeldonPayload) *proto.SeldonMessage {
+	if m, ok := p.GetPayload().(*proto.SeldonMessage); ok {
+		return m
+	}
+	return nil
 }

--- a/executor/api/grpc/seldon/server_test.go
+++ b/executor/api/grpc/seldon/server_test.go
@@ -2,14 +2,17 @@ package seldon
 
 import (
 	"context"
+	"net/url"
+	"testing"
+
 	"github.com/golang/protobuf/jsonpb"
 	. "github.com/onsi/gomega"
 	"github.com/seldonio/seldon-core/executor/api/grpc/seldon/proto"
 	"github.com/seldonio/seldon-core/executor/api/test"
-	"github.com/seldonio/seldon-core/operator/apis/machinelearning.seldon.io/v1"
-	"net/url"
-	"testing"
+	v1 "github.com/seldonio/seldon-core/operator/apis/machinelearning.seldon.io/v1"
 )
+
+const testSeldonPuid = "1"
 
 func TestPredict(t *testing.T) {
 	t.Logf("Started")

--- a/executor/api/grpc/server.go
+++ b/executor/api/grpc/server.go
@@ -4,7 +4,6 @@ import (
 	"context"
 	"math"
 	"strconv"
-	"strings"
 
 	"github.com/go-logr/logr"
 	guuid "github.com/google/uuid"
@@ -65,13 +64,13 @@ func CreateGrpcServer(spec *v1.PredictorSpec, deploymentName string, annotations
 	return grpcServer, nil
 }
 
-func CollectMetadata(ctx context.Context) map[string][]string {
-	if md, ok := metadata.FromIncomingContext(ctx); ok {
-		val := md.Get(payload.SeldonPUIDHeader)
+func CollectMetadata(ctx context.Context) metadata.MD {
+	if mdFromIncoming, ok := metadata.FromIncomingContext(ctx); ok {
+		val := mdFromIncoming.Get(payload.SeldonPUIDHeader)
 		if len(val) == 0 {
-			md.Set(payload.SeldonPUIDHeader, guuid.New().String())
+			mdFromIncoming.Set(payload.SeldonPUIDHeader, guuid.New().String())
 		}
-		return md
+		return mdFromIncoming
 	}
-	return map[string][]string{strings.ToLower(payload.SeldonPUIDHeader): []string{guuid.New().String()}}
+	return metadata.New(map[string]string{payload.SeldonPUIDHeader: guuid.New().String()})
 }

--- a/executor/api/grpc/server.go
+++ b/executor/api/grpc/server.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"math"
 	"strconv"
+	"strings"
 
 	"github.com/go-logr/logr"
 	guuid "github.com/google/uuid"
@@ -65,14 +66,12 @@ func CreateGrpcServer(spec *v1.PredictorSpec, deploymentName string, annotations
 }
 
 func CollectMetadata(ctx context.Context) map[string][]string {
-	md, ok := metadata.FromIncomingContext(ctx)
-	if ok {
+	if md, ok := metadata.FromIncomingContext(ctx); ok {
 		val := md.Get(payload.SeldonPUIDHeader)
 		if len(val) == 0 {
 			md.Set(payload.SeldonPUIDHeader, guuid.New().String())
 		}
 		return md
-	} else {
-		return map[string][]string{payload.SeldonPUIDHeader: []string{guuid.New().String()}}
 	}
+	return map[string][]string{strings.ToLower(payload.SeldonPUIDHeader): []string{guuid.New().String()}}
 }

--- a/executor/api/grpc/server_test.go
+++ b/executor/api/grpc/server_test.go
@@ -20,11 +20,11 @@ func TestAddPuid(t *testing.T) {
 
 func TestExistingPuid(t *testing.T) {
 	g := NewGomegaWithT(t)
-	guid := "1"
+	puid := "1"
 
-	ctx := metadata.NewIncomingContext(context.TODO(), metadata.New(map[string]string{payload.SeldonPUIDHeader: guid}))
+	ctx := metadata.NewIncomingContext(context.TODO(), metadata.New(map[string]string{payload.SeldonPUIDHeader: puid}))
 	meta := CollectMetadata(ctx)
 
 	g.Expect(meta.Get(payload.SeldonPUIDHeader)).NotTo(BeNil())
-	g.Expect(meta.Get(payload.SeldonPUIDHeader)[0]).To(Equal(guid))
+	g.Expect(meta.Get(payload.SeldonPUIDHeader)[0]).To(Equal(puid))
 }

--- a/executor/api/grpc/server_test.go
+++ b/executor/api/grpc/server_test.go
@@ -2,7 +2,6 @@ package grpc
 
 import (
 	"context"
-	"strings"
 	"testing"
 
 	. "github.com/onsi/gomega"
@@ -16,7 +15,7 @@ func TestAddPuid(t *testing.T) {
 	ctx := context.Background()
 	meta := CollectMetadata(ctx)
 
-	g.Expect(meta[strings.ToLower(payload.SeldonPUIDHeader)]).NotTo(BeNil())
+	g.Expect(meta.Get(payload.SeldonPUIDHeader)).NotTo(BeNil())
 }
 
 func TestExistingPuid(t *testing.T) {
@@ -26,6 +25,6 @@ func TestExistingPuid(t *testing.T) {
 	ctx := metadata.NewIncomingContext(context.TODO(), metadata.New(map[string]string{payload.SeldonPUIDHeader: guid}))
 	meta := CollectMetadata(ctx)
 
-	g.Expect(meta[strings.ToLower(payload.SeldonPUIDHeader)]).NotTo(BeNil())
-	g.Expect(meta[strings.ToLower(payload.SeldonPUIDHeader)][0]).To(Equal(guid))
+	g.Expect(meta.Get(payload.SeldonPUIDHeader)).NotTo(BeNil())
+	g.Expect(meta.Get(payload.SeldonPUIDHeader)[0]).To(Equal(guid))
 }

--- a/executor/api/grpc/server_test.go
+++ b/executor/api/grpc/server_test.go
@@ -2,11 +2,12 @@ package grpc
 
 import (
 	"context"
+	"strings"
+	"testing"
+
 	. "github.com/onsi/gomega"
 	"github.com/seldonio/seldon-core/executor/api/payload"
 	"google.golang.org/grpc/metadata"
-	"strings"
-	"testing"
 )
 
 func TestAddPuid(t *testing.T) {
@@ -15,7 +16,7 @@ func TestAddPuid(t *testing.T) {
 	ctx := context.Background()
 	meta := CollectMetadata(ctx)
 
-	g.Expect(meta[payload.SeldonPUIDHeader]).NotTo(BeNil())
+	g.Expect(meta[strings.ToLower(payload.SeldonPUIDHeader)]).NotTo(BeNil())
 }
 
 func TestExistingPuid(t *testing.T) {

--- a/executor/api/grpc/tensorflow/server.go
+++ b/executor/api/grpc/tensorflow/server.go
@@ -2,6 +2,9 @@ package tensorflow
 
 import (
 	"context"
+	"net/url"
+	"strings"
+
 	"github.com/go-logr/logr"
 	"github.com/golang/protobuf/proto"
 	"github.com/pkg/errors"
@@ -10,8 +13,7 @@ import (
 	"github.com/seldonio/seldon-core/executor/api/payload"
 	"github.com/seldonio/seldon-core/executor/predictor"
 	"github.com/seldonio/seldon-core/executor/proto/tensorflow/serving"
-	"github.com/seldonio/seldon-core/operator/apis/machinelearning.seldon.io/v1"
-	"net/url"
+	v1 "github.com/seldonio/seldon-core/operator/apis/machinelearning.seldon.io/v1"
 	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
 )
 
@@ -34,7 +36,9 @@ func NewGrpcTensorflowServer(predictor *v1.PredictorSpec, client client.SeldonAp
 }
 
 func (g *GrpcTensorflowServer) execute(ctx context.Context, req proto.Message, method string) (payload.SeldonPayload, error) {
-	seldonPredictorProcess := predictor.NewPredictorProcess(ctx, g.Client, logf.Log.WithName(method), g.ServerUrl, g.Namespace, grpc.CollectMetadata(ctx))
+	md := grpc.CollectMetadata(ctx)
+	ctx = context.WithValue(ctx, payload.SeldonPUIDHeader, md[strings.ToLower(payload.SeldonPUIDHeader)][0])
+	seldonPredictorProcess := predictor.NewPredictorProcess(ctx, g.Client, logf.Log.WithName(method), g.ServerUrl, g.Namespace, md)
 	reqPayload := payload.ProtoPayload{Msg: req}
 	return seldonPredictorProcess.Predict(g.predictor.Graph, &reqPayload)
 }

--- a/executor/api/grpc/tensorflow/server.go
+++ b/executor/api/grpc/tensorflow/server.go
@@ -3,7 +3,6 @@ package tensorflow
 import (
 	"context"
 	"net/url"
-	"strings"
 
 	"github.com/go-logr/logr"
 	"github.com/golang/protobuf/proto"
@@ -37,7 +36,7 @@ func NewGrpcTensorflowServer(predictor *v1.PredictorSpec, client client.SeldonAp
 
 func (g *GrpcTensorflowServer) execute(ctx context.Context, req proto.Message, method string) (payload.SeldonPayload, error) {
 	md := grpc.CollectMetadata(ctx)
-	ctx = context.WithValue(ctx, payload.SeldonPUIDHeader, md[strings.ToLower(payload.SeldonPUIDHeader)][0])
+	ctx = context.WithValue(ctx, payload.SeldonPUIDHeader, md.Get(payload.SeldonPUIDHeader)[0])
 	seldonPredictorProcess := predictor.NewPredictorProcess(ctx, g.Client, logf.Log.WithName(method), g.ServerUrl, g.Namespace, md)
 	reqPayload := payload.ProtoPayload{Msg: req}
 	return seldonPredictorProcess.Predict(g.predictor.Graph, &reqPayload)

--- a/executor/api/rest/server.go
+++ b/executor/api/rest/server.go
@@ -3,6 +3,10 @@ package rest
 import (
 	"context"
 	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
 	"github.com/go-logr/logr"
 	guuid "github.com/google/uuid"
 	"github.com/gorilla/mux"
@@ -15,10 +19,7 @@ import (
 	"github.com/seldonio/seldon-core/executor/api/metric"
 	"github.com/seldonio/seldon-core/executor/api/payload"
 	"github.com/seldonio/seldon-core/executor/predictor"
-	"github.com/seldonio/seldon-core/operator/apis/machinelearning.seldon.io/v1"
-	"io/ioutil"
-	"net/http"
-	"net/url"
+	v1 "github.com/seldonio/seldon-core/operator/apis/machinelearning.seldon.io/v1"
 	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
 )
 

--- a/executor/api/rest/server_test.go
+++ b/executor/api/rest/server_test.go
@@ -1,6 +1,14 @@
 package rest
 
 import (
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strconv"
+	"strings"
+	"testing"
+
 	guuid "github.com/google/uuid"
 	. "github.com/onsi/gomega"
 	"github.com/prometheus/common/expfmt"
@@ -9,13 +17,6 @@ import (
 	"github.com/seldonio/seldon-core/executor/api/payload"
 	"github.com/seldonio/seldon-core/executor/api/test"
 	v1 "github.com/seldonio/seldon-core/operator/apis/machinelearning.seldon.io/v1"
-	"io/ioutil"
-	"net/http"
-	"net/http/httptest"
-	"net/url"
-	"strconv"
-	"strings"
-	"testing"
 )
 
 const (
@@ -266,7 +267,6 @@ func TestModelWithServer(t *testing.T) {
 	r.Router.ServeHTTP(res, req)
 	g.Expect(res.Code).To(Equal(200))
 	g.Expect(called).To(Equal(true))
-
 }
 
 func TestServerMetrics(t *testing.T) {

--- a/executor/predictor/predictor_process.go
+++ b/executor/predictor/predictor_process.go
@@ -274,7 +274,10 @@ func (p *PredictorProcess) logPayload(nodeName string, logger *v1.Logger, reqTyp
 func (p *PredictorProcess) Predict(node *v1.PredictiveUnit, msg payload.SeldonPayload) (payload.SeldonPayload, error) {
 	//Log Request
 	if node.Logger != nil && (node.Logger.Mode == v1.LogRequest || node.Logger.Mode == v1.LogAll) {
-		p.logPayload(node.Name, node.Logger, payloadLogger.InferenceRequest, msg)
+		err := p.logPayload(node.Name, node.Logger, payloadLogger.InferenceRequest, msg)
+		if err != nil {
+			return nil, err
+		}
 	}
 	tmsg, err := p.transformInput(node, msg)
 	if err != nil {
@@ -287,7 +290,10 @@ func (p *PredictorProcess) Predict(node *v1.PredictiveUnit, msg payload.SeldonPa
 	response, err := p.transformOutput(node, cmsg)
 	// Log Response
 	if err == nil && node.Logger != nil && (node.Logger.Mode == v1.LogResponse || node.Logger.Mode == v1.LogAll) {
-		p.logPayload(node.Name, node.Logger, payloadLogger.InferenceResponse, response)
+		err := p.logPayload(node.Name, node.Logger, payloadLogger.InferenceResponse, response)
+		if err != nil {
+			return nil, err
+		}
 	}
 	return response, err
 }

--- a/executor/predictor/predictor_process.go
+++ b/executor/predictor/predictor_process.go
@@ -243,7 +243,7 @@ func (p *PredictorProcess) getLogUrl(logger *v1.Logger) (*url.URL, error) {
 	}
 }
 
-func (p *PredictorProcess) logPayload(nodeName string, logger *v1.Logger, reqType payloadLogger.LogRequestType, msg payload.SeldonPayload) error {
+func (p *PredictorProcess) logPayload(nodeName string, logger *v1.Logger, reqType payloadLogger.LogRequestType, msg payload.SeldonPayload, puid string) error {
 	data, err := msg.GetBytes()
 	if err != nil {
 		return err
@@ -251,12 +251,6 @@ func (p *PredictorProcess) logPayload(nodeName string, logger *v1.Logger, reqTyp
 	logUrl, err := p.getLogUrl(logger)
 	if err != nil {
 		return err
-	}
-	var reqId string
-	if r, ok := p.Ctx.Value(payload.SeldonPUIDHeader).(string); ok {
-		reqId = r
-	} else {
-		return fmt.Errorf("context value Seldon PUID Header is nil: interface to string conversion failed")
 	}
 	payloadLogger.QueueLogRequest(payloadLogger.LogRequest{
 		Url:         logUrl,
@@ -266,15 +260,27 @@ func (p *PredictorProcess) logPayload(nodeName string, logger *v1.Logger, reqTyp
 		Id:          guuid.New().String(),
 		SourceUri:   p.ServerUrl,
 		ModelId:     nodeName,
-		RequestId:   reqId,
+		RequestId:   puid,
 	})
 	return nil
 }
 
+func (p *PredictorProcess) getPUIDHeader() (string, error) {
+	// Check request ID is not nil
+	if puid, ok := p.Ctx.Value(payload.SeldonPUIDHeader).(string); ok {
+		return puid, nil
+	}
+	return "", fmt.Errorf("context value Seldon PUID Header is nil: interface to string conversion failed")
+}
+
 func (p *PredictorProcess) Predict(node *v1.PredictiveUnit, msg payload.SeldonPayload) (payload.SeldonPayload, error) {
+	puid, err := p.getPUIDHeader()
+	if err != nil {
+		return nil, err
+	}
 	//Log Request
 	if node.Logger != nil && (node.Logger.Mode == v1.LogRequest || node.Logger.Mode == v1.LogAll) {
-		err := p.logPayload(node.Name, node.Logger, payloadLogger.InferenceRequest, msg)
+		err := p.logPayload(node.Name, node.Logger, payloadLogger.InferenceRequest, msg, puid)
 		if err != nil {
 			return nil, err
 		}
@@ -290,7 +296,7 @@ func (p *PredictorProcess) Predict(node *v1.PredictiveUnit, msg payload.SeldonPa
 	response, err := p.transformOutput(node, cmsg)
 	// Log Response
 	if err == nil && node.Logger != nil && (node.Logger.Mode == v1.LogResponse || node.Logger.Mode == v1.LogAll) {
-		err := p.logPayload(node.Name, node.Logger, payloadLogger.InferenceResponse, response)
+		err := p.logPayload(node.Name, node.Logger, payloadLogger.InferenceResponse, response, puid)
 		if err != nil {
 			return nil, err
 		}

--- a/executor/predictor/predictor_process.go
+++ b/executor/predictor/predictor_process.go
@@ -14,6 +14,8 @@ import (
 	v1 "github.com/seldonio/seldon-core/operator/apis/machinelearning.seldon.io/v1"
 )
 
+const NilPUIDError = "context value for Seldon PUID Header is nil"
+
 type PredictorProcess struct {
 	Ctx       context.Context
 	Client    client.SeldonApiClient
@@ -270,7 +272,7 @@ func (p *PredictorProcess) getPUIDHeader() (string, error) {
 	if puid, ok := p.Ctx.Value(payload.SeldonPUIDHeader).(string); ok {
 		return puid, nil
 	}
-	return "", fmt.Errorf("context value Seldon PUID Header is nil: interface to string conversion failed")
+	return "", fmt.Errorf(NilPUIDError)
 }
 
 func (p *PredictorProcess) Predict(node *v1.PredictiveUnit, msg payload.SeldonPayload) (payload.SeldonPayload, error) {

--- a/executor/predictor/predictor_process_test.go
+++ b/executor/predictor/predictor_process_test.go
@@ -472,5 +472,5 @@ func TestPredictNilPUIDError(t *testing.T) {
 	graph := &v1.PredictiveUnit{}
 	_, err := createPredictorProcessWithoutPUIDInContext(t).Predict(graph, createPredictPayload(g))
 	g.Expect(err).NotTo(BeNil())
-	g.Expect(err.Error()).Should(Equal("context value Seldon PUID Header is nil: interface to string conversion failed"))
+	g.Expect(err.Error()).Should(Equal(NilPUIDError))
 }

--- a/executor/predictor/predictor_process_test.go
+++ b/executor/predictor/predictor_process_test.go
@@ -4,6 +4,11 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+
 	"github.com/golang/protobuf/jsonpb"
 	. "github.com/onsi/gomega"
 	"github.com/seldonio/seldon-core/executor/api/grpc"
@@ -12,11 +17,7 @@ import (
 	"github.com/seldonio/seldon-core/executor/api/test"
 	"github.com/seldonio/seldon-core/executor/logger"
 	v1 "github.com/seldonio/seldon-core/operator/apis/machinelearning.seldon.io/v1"
-	"net/http"
-	"net/http/httptest"
-	"net/url"
 	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
-	"testing"
 )
 
 const (
@@ -47,6 +48,13 @@ func createPredictorProcessWithError(t *testing.T, errMethod *v1.PredictiveUnitM
 	url, _ := url.Parse(testSourceUrl)
 	ctx := context.WithValue(context.TODO(), payload.SeldonPUIDHeader, testSeldonPuid)
 	pp := NewPredictorProcess(ctx, &test.SeldonMessageTestClient{ErrMethod: errMethod, Err: err, ErrPayload: errPayload}, logf.Log.WithName("SeldonMessageRestClient"), url, "default", map[string][]string{})
+	return &pp
+}
+
+func createPredictorProcessWithoutPUID(t *testing.T) *PredictorProcess {
+	url, _ := url.Parse(testSourceUrl)
+	ctx := context.WithValue(context.TODO(), payload.SeldonPUIDHeader, nil)
+	pp := NewPredictorProcess(ctx, &test.SeldonMessageTestClient{}, logf.Log.WithName("SeldonMessageRestClient"), url, "default", map[string][]string{testCustomMetaKey: []string{testCustomMetaValue}})
 	return &pp
 }
 
@@ -454,4 +462,86 @@ func TestModelWithLogResponses(t *testing.T) {
 	g.Expect(smRes.GetData().GetNdarray().Values[0].GetNumberValue()).Should(Equal(1.1))
 	g.Expect(smRes.GetData().GetNdarray().Values[1].GetNumberValue()).Should(Equal(2.0))
 	g.Eventually(func() bool { return logged }).Should(Equal(true))
+}
+
+func TestModelWithLogRequestsNilPUIDError(t *testing.T) {
+	t.Logf("Started")
+	g := NewGomegaWithT(t)
+	modelName := "foo"
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		//g.Expect(r.Header.Get(logger.CloudEventsIdHeader)).Should(Equal(testEventId))
+		g.Expect(r.Header.Get(logger.CloudEventsTypeHeader)).To(Equal(logger.CEInferenceRequest))
+		g.Expect(r.Header.Get(logger.CloudEventsTypeSource)).To(Equal(testSourceUrl))
+		g.Expect(r.Header.Get(modelIdHeaderName)).To(Equal(modelName))
+		g.Expect(r.Header.Get(contentTypeHeaderName)).To(Equal(grpc.ProtobufContentType))
+		g.Expect(r.Header.Get(requestIdHeaderName)).To(Equal(testSeldonPuid))
+		w.Write([]byte(""))
+		fmt.Printf("%+v\n", r.Header)
+		fmt.Printf("%+v\n", r.Body)
+	})
+	server := httptest.NewServer(handler)
+	defer server.Close()
+
+	logf.SetLogger(logf.ZapLogger(false))
+	log := logf.Log.WithName("entrypoint")
+	logger.StartDispatcher(1, log, "", "", "")
+
+	model := v1.MODEL
+	graph := &v1.PredictiveUnit{
+		Name: modelName,
+		Type: &model,
+		Endpoint: &v1.Endpoint{
+			ServiceHost: "foo",
+			ServicePort: 9000,
+			Type:        v1.REST,
+		},
+		Logger: &v1.Logger{
+			Mode: v1.LogRequest,
+			Url:  &server.URL,
+		},
+	}
+
+	_, err := createPredictorProcessWithoutPUID(t).Predict(graph, createPredictPayload(g))
+	g.Expect(err).NotTo(BeNil())
+	g.Expect(err.Error()).Should(Equal("context value Seldon PUID Header is nil: interface to string conversion failed"))
+}
+
+func TestModelWithLogResponsesNilPUIDError(t *testing.T) {
+	t.Logf("Started")
+	g := NewGomegaWithT(t)
+	modelName := "foo"
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		//g.Expect(r.Header.Get(logger.CloudEventsIdHeader)).Should(Equal(testEventId))
+		g.Expect(r.Header.Get(logger.CloudEventsTypeHeader)).To(Equal(logger.CEInferenceResponse))
+		g.Expect(r.Header.Get(logger.CloudEventsTypeSource)).To(Equal(testSourceUrl))
+		g.Expect(r.Header.Get(modelIdHeaderName)).To(Equal(modelName))
+		g.Expect(r.Header.Get(contentTypeHeaderName)).To(Equal(grpc.ProtobufContentType))
+		g.Expect(r.Header.Get(requestIdHeaderName)).To(Equal(testSeldonPuid))
+		w.Write([]byte(""))
+	})
+	server := httptest.NewServer(handler)
+	defer server.Close()
+
+	logf.SetLogger(logf.ZapLogger(false))
+	log := logf.Log.WithName("entrypoint")
+	logger.StartDispatcher(1, log, "", "", "")
+
+	model := v1.MODEL
+	graph := &v1.PredictiveUnit{
+		Name: modelName,
+		Type: &model,
+		Endpoint: &v1.Endpoint{
+			ServiceHost: "foo",
+			ServicePort: 9000,
+			Type:        v1.REST,
+		},
+		Logger: &v1.Logger{
+			Mode: v1.LogResponse,
+			Url:  &server.URL,
+		},
+	}
+
+	_, err := createPredictorProcessWithoutPUID(t).Predict(graph, createPredictPayload(g))
+	g.Expect(err).NotTo(BeNil())
+	g.Expect(err.Error()).Should(Equal("context value Seldon PUID Header is nil: interface to string conversion failed"))
 }

--- a/executor/predictor/predictor_process_test.go
+++ b/executor/predictor/predictor_process_test.go
@@ -51,9 +51,9 @@ func createPredictorProcessWithError(t *testing.T, errMethod *v1.PredictiveUnitM
 	return &pp
 }
 
-func createPredictorProcessWithoutPUID(t *testing.T) *PredictorProcess {
+func createPredictorProcessWithoutPUIDInContext(t *testing.T) *PredictorProcess {
 	url, _ := url.Parse(testSourceUrl)
-	ctx := context.WithValue(context.TODO(), payload.SeldonPUIDHeader, nil)
+	ctx := context.TODO()
 	pp := NewPredictorProcess(ctx, &test.SeldonMessageTestClient{}, logf.Log.WithName("SeldonMessageRestClient"), url, "default", map[string][]string{testCustomMetaKey: []string{testCustomMetaValue}})
 	return &pp
 }
@@ -378,7 +378,6 @@ func TestModelWithLogRequests(t *testing.T) {
 	modelName := "foo"
 	logged := false
 	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		//g.Expect(r.Header.Get(logger.CloudEventsIdHeader)).Should(Equal(testEventId))
 		g.Expect(r.Header.Get(logger.CloudEventsTypeHeader)).To(Equal(logger.CEInferenceRequest))
 		g.Expect(r.Header.Get(logger.CloudEventsTypeSource)).To(Equal(testSourceUrl))
 		g.Expect(r.Header.Get(modelIdHeaderName)).To(Equal(modelName))
@@ -425,7 +424,6 @@ func TestModelWithLogResponses(t *testing.T) {
 	modelName := "foo"
 	logged := false
 	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		//g.Expect(r.Header.Get(logger.CloudEventsIdHeader)).Should(Equal(testEventId))
 		g.Expect(r.Header.Get(logger.CloudEventsTypeHeader)).To(Equal(logger.CEInferenceResponse))
 		g.Expect(r.Header.Get(logger.CloudEventsTypeSource)).To(Equal(testSourceUrl))
 		g.Expect(r.Header.Get(modelIdHeaderName)).To(Equal(modelName))
@@ -464,84 +462,15 @@ func TestModelWithLogResponses(t *testing.T) {
 	g.Eventually(func() bool { return logged }).Should(Equal(true))
 }
 
-func TestModelWithLogRequestsNilPUIDError(t *testing.T) {
+func TestPredictNilPUIDError(t *testing.T) {
 	t.Logf("Started")
 	g := NewGomegaWithT(t)
-	modelName := "foo"
 	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		//g.Expect(r.Header.Get(logger.CloudEventsIdHeader)).Should(Equal(testEventId))
-		g.Expect(r.Header.Get(logger.CloudEventsTypeHeader)).To(Equal(logger.CEInferenceRequest))
-		g.Expect(r.Header.Get(logger.CloudEventsTypeSource)).To(Equal(testSourceUrl))
-		g.Expect(r.Header.Get(modelIdHeaderName)).To(Equal(modelName))
-		g.Expect(r.Header.Get(contentTypeHeaderName)).To(Equal(grpc.ProtobufContentType))
-		g.Expect(r.Header.Get(requestIdHeaderName)).To(Equal(testSeldonPuid))
-		w.Write([]byte(""))
-		fmt.Printf("%+v\n", r.Header)
-		fmt.Printf("%+v\n", r.Body)
 	})
 	server := httptest.NewServer(handler)
 	defer server.Close()
-
-	logf.SetLogger(logf.ZapLogger(false))
-	log := logf.Log.WithName("entrypoint")
-	logger.StartDispatcher(1, log, "", "", "")
-
-	model := v1.MODEL
-	graph := &v1.PredictiveUnit{
-		Name: modelName,
-		Type: &model,
-		Endpoint: &v1.Endpoint{
-			ServiceHost: "foo",
-			ServicePort: 9000,
-			Type:        v1.REST,
-		},
-		Logger: &v1.Logger{
-			Mode: v1.LogRequest,
-			Url:  &server.URL,
-		},
-	}
-
-	_, err := createPredictorProcessWithoutPUID(t).Predict(graph, createPredictPayload(g))
-	g.Expect(err).NotTo(BeNil())
-	g.Expect(err.Error()).Should(Equal("context value Seldon PUID Header is nil: interface to string conversion failed"))
-}
-
-func TestModelWithLogResponsesNilPUIDError(t *testing.T) {
-	t.Logf("Started")
-	g := NewGomegaWithT(t)
-	modelName := "foo"
-	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		//g.Expect(r.Header.Get(logger.CloudEventsIdHeader)).Should(Equal(testEventId))
-		g.Expect(r.Header.Get(logger.CloudEventsTypeHeader)).To(Equal(logger.CEInferenceResponse))
-		g.Expect(r.Header.Get(logger.CloudEventsTypeSource)).To(Equal(testSourceUrl))
-		g.Expect(r.Header.Get(modelIdHeaderName)).To(Equal(modelName))
-		g.Expect(r.Header.Get(contentTypeHeaderName)).To(Equal(grpc.ProtobufContentType))
-		g.Expect(r.Header.Get(requestIdHeaderName)).To(Equal(testSeldonPuid))
-		w.Write([]byte(""))
-	})
-	server := httptest.NewServer(handler)
-	defer server.Close()
-
-	logf.SetLogger(logf.ZapLogger(false))
-	log := logf.Log.WithName("entrypoint")
-	logger.StartDispatcher(1, log, "", "", "")
-
-	model := v1.MODEL
-	graph := &v1.PredictiveUnit{
-		Name: modelName,
-		Type: &model,
-		Endpoint: &v1.Endpoint{
-			ServiceHost: "foo",
-			ServicePort: 9000,
-			Type:        v1.REST,
-		},
-		Logger: &v1.Logger{
-			Mode: v1.LogResponse,
-			Url:  &server.URL,
-		},
-	}
-
-	_, err := createPredictorProcessWithoutPUID(t).Predict(graph, createPredictPayload(g))
+	graph := &v1.PredictiveUnit{}
+	_, err := createPredictorProcessWithoutPUIDInContext(t).Predict(graph, createPredictPayload(g))
 	g.Expect(err).NotTo(BeNil())
 	g.Expect(err.Error()).Should(Equal("context value Seldon PUID Header is nil: interface to string conversion failed"))
 }

